crates/ruff_python_formatter/src/other/f_string_element.rs

```diff
@@ -0,0 +1,264 @@
+                [
+                    text(&debug_text.leading),
+                    suppressed_node(&**expression),
+                    text(&debug_text.trailing),
+                ]
+            )?;
+
+            // Even if debug text is present, any whitespace between the
+            // conversion flag and the format spec doesn't need to be preserved.
+            match conversion {
+                ConversionFlag::Str => text("!s").fmt(f)?,
+                ConversionFlag::Ascii => text("!a").fmt(f)?,
+                ConversionFlag::Repr => text("!r").fmt(f)?,
+                ConversionFlag::None => (),
+            }
+
+            if let Some(format_spec) = format_spec.as_deref() {
+                write!(f, [token(":"), suppressed_node(format_spec)])?;
+            }
+
+            token("}").fmt(f)
+        } else {
+            let dangling_item_comments = comments.dangling(self.element);
+            let (dangling_open_parentheses_comments, trailing_format_spec_comments) =
+                dangling_item_comments.split_at(
+                    dangling_item_comments
+                        .partition_point(|comment| comment.start() < expression.start()),
+                );
+
+            let item = format_with(|f| {
+                let line_break_or_space = match expression.as_ref() {
+                    // If an expression starts with a `{`, we need to add a space before the
+                    // curly brace to avoid turning it into a literal curly with `{{`.
+                    //
+                    // For example,
+                    // ```python
+                    // f"{ {'x': 1, 'y': 2} }"
+                    // #  ^                ^
+                    // ```
+                    //
+                    // We need to preserve the space highlighted by `^`.
+                    Expr::Dict(_) | Expr::DictComp(_) | Expr::Set(_) | Expr::SetComp(_) => {
+                        Some(soft_line_break_or_space())
+                    }
+                    _ => None,
+                };
+
+                let f = &mut WithExprLocation::new(
+                    ExpressionLocation::InsideFString(self.context.quotes()),
+                    f,
+                );
+
+                line_break_or_space.fmt(f)?;
+
+                // If we're going to remove the soft line breaks, then there's a chance
+                // that there will be trailing commas in the formatted expression.
+                // Currently, it's difficult to remove that conditionally, because (TODO)
+                // So, we'll manually format the expression with the maximum line width
+                // and disabling the magic trailing comma. This is expensive so we've
+                // implemented some heuristics to avoid this in some cases.
+                if self.context.should_remove_soft_line_breaks()
+                    && !matches!(
+                        &**expression,
+                        Expr::BooleanLiteral(_)
```

Comment:
Nit: Can we avoid this as well if we have a simple attribute chain `a.b.c`?

Reply:
I'm using the preorder visitor which should be more robust.